File: NeetCode/Heap/Single_Threaded_CPU/Single_Threaded_CPU.cpp
// Single_Threaded_CPU.cpp
//
// Compile locally with:
// clang++ Single_Threaded_CPU.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> getOrder(vector<vector<int>> &tasks)
    {
        vector<vector<int>> taskList;
        for (int idx = 0; idx < tasks.size(); idx++)
        {
            int startTime = tasks[idx][0];
            int duration = tasks[idx][1];
            taskList.push_back({startTime, duration, idx});
        }

        sort(taskList.begin(), taskList.end());
        int n = taskList.size();
        int time(0), index(0), ptime(0), orgIndex(0);

        priority_queue<pair<int, int>, vector<pair<int, int>>,
                       greater<pair<int, int>>>
            available;
        vector<int> result;

        while (index < n || !available.empty())
        {
            if (available.empty() && time < taskList[index][0])
            {
                time = taskList[index][0];
            }

            while (index < n && taskList[index][0] <= time)
            {
                ptime = taskList[index][1];
                orgIndex = taskList[index][2];
                available.push({ptime, orgIndex});
                index++;
            }

            auto tmp = available.top();
            available.pop();
            result.push_back(tmp.second);
            time += tmp.first;
        }
        return result;
    }
};

int main()
{
    cout << "Running Single_Threaded_CPU.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: NeetCode/Heap/Single_Threaded_CPU/single__threaded__c_p_u.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def getOrder(self, tasks: List[List[int]]) -> List[int]:
        taskList = []
        for index, (start_time, duration) in enumerate(tasks):
            taskList.append((start_time, duration, index))

        taskList.sort()

        n = len(tasks)
        time = 0
        index = 0

        available = []
        results = []

        while index < n or available:

            if not available and time < taskList[index][0]:
                time = taskList[index][0]

            while index < n and taskList[index][0] <= time:

                _, ptime, org_index = taskList[index]
                heapq.heappush(available, (ptime, org_index))
                index += 1

            ptime, org_index = heapq.heappop(available)

            results.append(org_index)
            time += ptime

        return results


if __name__ == "__main__":
    sol = Solution()
    tasks = [[5, 2], [4, 4], [4, 1], [2, 1], [3, 3]]
    print(sol.getOrder(tasks))
    print("Running Solution...")
